solutions/devsprint-matheus-reis-2/DeliveryAppChallenge/Service/DeliveryApi.swift

```diff
@@ -7,25 +7,63 @@
 
 import Foundation
 
-struct DeliveryApi {
+enum UrlName {
+    static let restaurantList: String = "home_restaurant_list.json"
+}
 
-    func fetchRestaurants(_ completion: ([String]) -> Void) {
+protocol DeliveryApiProtocol {
+    func fetchRestaurants(_ name: String, _ completion: @escaping (Result<[Restaurant], DeliveryApiError>) -> Void)
+    func searchAddresses(_ completion: ([String]) -> Void)
+    func fetchRestaurantDetails(_ completion: (String) -> Void)
+    func fetchMenuItem(_ completion: (String) -> Void)
+}
 
-        completion(["Restaurant 1", "Restaurant 2", "Restaurant 3"])
+struct DeliveryApi: DeliveryApiProtocol {
+    
+    func fetchRestaurants(_ name: String, _ completion: @escaping (Result<[Restaurant], DeliveryApiError>) -> Void) {
+        self.request(name, completion: completion   )
     }
-
+    
     func searchAddresses(_ completion: ([String]) -> Void) {
-
-        completion(["Address 1", "Address 2", "Address 3"])
+        
     }
-
+    
     func fetchRestaurantDetails(_ completion: (String) -> Void) {
-
-        completion("Restaurant Details")
+        
     }
-
+    
     func fetchMenuItem(_ completion: (String) -> Void) {
-
-        completion("Menu Item")
+        
+    }
+    
+    private func request<T: Decodable>(_ name: String, completion: @escaping (Result<T, DeliveryApiError>) -> Void){
+        guard let  url = URL(string: "https://raw.githubusercontent.com/devpass-tech/challenge-mvvm-delivery/main/api/\(name)") else { return }
```

Comment:
Você criou o erro `invalidURL` no enum mas não usou aqui. Imagino que tenha esquecido.

```suggestion
        guard let  url = URL(string: "https://raw.githubusercontent.com/devpass-tech/challenge-mvvm-delivery/main/api/\(name)") else {
            completion(.failure(.invalidURL))
            return
        }
```